File: src/net/minecraft/client/gui/screen/option/VideoOptionsScreen.cpp
#include "net/minecraft/client/gui/screen/option/VideoOptionsScreen.hpp"
#include "net/minecraft/client/Minecraft.hpp"
#include "net/minecraft/client/gui/layout/OptionsLayout.hpp"
#include "net/minecraft/client/gui/layout/ScreenLayout.hpp"
#include "net/minecraft/client/gui/screen/option/AnimationSettingsScreen.hpp"
#include "net/minecraft/client/gui/screen/option/DetailSettingsScreen.hpp"
#include "net/minecraft/client/gui/screen/option/PerformanceSettingsScreen.hpp"
#include "net/minecraft/client/gui/screen/option/QualitySettingsScreen.hpp"
#include "net/minecraft/client/gui/screen/option/WorldSettingsScreen.hpp"
#include "net/minecraft/client/resource/language/I18n.hpp"
namespace net::minecraft::client::gui::screen::option {
VideoOptionsScreen::VideoOptionsScreen(ParentFactory parentFactory, client_option::GameOptions* options)
    : parentFactory_(std::move(parentFactory)), options_(options) {
}
void VideoOptionsScreen::init() {
  title_ = resource::language::I18n::getTranslation("options.videoTitle");
  buttons_.clear();
  if(options_ == nullptr || minecraft() == nullptr) {
    return;
  }
  const int startY = height() / 6;
  const int x1 = layout::optionsGridX(width(), 0);
  const int x2 = layout::optionsGridX(width(), 1);
  const auto returnHere = [factory = parentFactory_, options = options_]() {
    return std::make_unique<VideoOptionsScreen>(factory, options);
  };
  addActionButton(x1, startY, 310, layout::kDefaultButtonHeight, "Quality...", [this, returnHere] {
    options_->save();
    navigateTo([returnHere, opts = options_]() { return makeQualitySettingsScreen(returnHere, opts); });
  });
  addActionButton(
      x1, startY + layout::kRowSpacing, 150, layout::kDefaultButtonHeight, "Performance...", [this, returnHere] {
        options_->save();
        navigateTo([returnHere, opts = options_]() { return makePerformanceSettingsScreen(returnHere, opts); });
      });
  addActionButton(
      x2, startY + layout::kRowSpacing, 150, layout::kDefaultButtonHeight, "Details...", [this, returnHere] {
        options_->save();
        navigateTo([returnHere, opts = options_]() { return makeDetailSettingsScreen(returnHere, opts); });
      });
  addActionButton(
      x1, startY + layout::kRowSpacing * 2, 150, layout::kDefaultButtonHeight, "Animations...", [this, returnHere] {
        options_->save();
        navigateTo([returnHere, opts = options_]() { return makeAnimationSettingsScreen(returnHere, opts); });
      });
  addActionButton(
      x2, startY + layout::kRowSpacing * 2, 150, layout::kDefaultButtonHeight, "World...", [this, returnHere] {
        options_->save();
        navigateTo(
            [returnHere, opts = options_]() { return std::make_unique<WorldSettingsScreen>(returnHere, opts); });
      });
  addCenteredActionButton(
      startY + layout::kRowSpacing * 3, resource::language::I18n::getTranslation("gui.done"), [this] {
        if(options_ == nullptr || minecraft() == nullptr) {
          return;
        }
        options_->save();
        navigateTo(parentFactory_);
      });
}
void VideoOptionsScreen::render(int mouseX, int mouseY, float tickDelta) {
  renderBackground();
  if(textRenderer() != nullptr) {
    drawCenteredTextWithShadow(*textRenderer(), title_, width() / 2, 20, 0xFFFFFF);
  }
  Screen::render(mouseX, mouseY, tickDelta);
}
} // namespace net::minecraft::client::gui::screen::option

File: mcp/src/net/minecraft/client/gui/screen/option/VideoOptionsScreen.java
package net.minecraft.client.gui.screen.option;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.widget.ButtonWidget;
import net.minecraft.client.option.GameOptions;
import net.minecraft.client.resource.language.TranslationStorage;

@Environment(value=EnvType.CLIENT)
public class VideoOptionsScreen
extends Screen {
    private Screen parent;
    protected String title = "Video Settings";
    private GameOptions options;

    public VideoOptionsScreen(Screen parent, GameOptions options) {
        this.parent = parent;
        this.options = options;
    }

    public void init() {
        TranslationStorage ts = TranslationStorage.getInstance();
        this.title = ts.get("options.videoTitle");
        int startY = this.height / 6;
        int x1 = this.width / 2 - 155;
        int x2 = this.width / 2 + 5;

        this.buttons.add(new WideButtonWidget(100, this.width / 2 - 155, startY, 310, "Quality..."));
        this.buttons.add(new ButtonWidget(101, x1, startY + 24, 150, 20, "Performance..."));
        this.buttons.add(new ButtonWidget(102, x2, startY + 24, 150, 20, "Details..."));
        this.buttons.add(new ButtonWidget(103, x1, startY + 48, 150, 20, "Animations..."));
        this.buttons.add(new ButtonWidget(104, x2, startY + 48, 150, 20, "World..."));
        this.buttons.add(new ButtonWidget(105, x1, startY + 72, 150, 20, "3D Settings..."));
        this.buttons.add(new ButtonWidget(200, this.width / 2 - 100, startY + 96, ts.get("gui.done")));
    }

    protected void buttonClicked(ButtonWidget button) {
        if (!button.active) return;
        if (button.id == 100) { this.minecraft.options.save(); this.minecraft.setScreen(new QualitySettingsScreen(this, this.options)); }
        if (button.id == 101) { this.minecraft.options.save(); this.minecraft.setScreen(new PerformanceSettingsScreen(this, this.options)); }
        if (button.id == 102) { this.minecraft.options.save(); this.minecraft.setScreen(new DetailSettingsScreen(this, this.options)); }
        if (button.id == 103) { this.minecraft.options.save(); this.minecraft.setScreen(new AnimationSettingsScreen(this, this.options)); }
        if (button.id == 104) { this.minecraft.options.save(); this.minecraft.setScreen(new WorldSettingsScreen(this, this.options)); }
        if (button.id == 105) { this.minecraft.options.save(); this.minecraft.setScreen(new StereoSettingsScreen(this, this.options)); }
        if (button.id == 200) { this.minecraft.options.save(); this.minecraft.setScreen(this.parent); }
    }

    public void render(int mouseX, int mouseY, float delta) {
        this.renderBackground();
        this.drawCenteredTextWithShadow(this.textRenderer, this.title, this.width / 2, 20, 0xFFFFFF);
        super.render(mouseX, mouseY, delta);
    }

    static class WideButtonWidget extends ButtonWidget {
        public WideButtonWidget(int id, int x, int y, int width, String text) {
            super(id, x, y, width, 20, text);
        }
    }
}
